recirq/qml_lfe/learn_dynamics_q.py

```diff
@@ -0,0 +1,238 @@
+
+    # Add tsym or scrambling block on System A qubits.
+    ret_circuit += circuit_blocks.block_1d_circuit(
+        [qbpair[0] for qbpair in qubit_pairs], depth, inter_gen, random_source
+    )
+
+    # Swap system A and system B qubits.
+    ret_circuit += [circuit_blocks.swap_block(pair) for pair in qubit_pairs]
+
+    # Again run tsym or scrambling block on newly swapped qubits.
+    ret_circuit += circuit_blocks.block_1d_circuit(
+        [qbpair[0] for qbpair in qubit_pairs], depth, inter_gen, random_source
+    )
+
+    # Un bell pair.
+    ret_circuit += [circuit_blocks.un_bell_pair_block(pair) for pair in qubit_pairs]
+
+    # Merge single qubit gates together and add measurements.
+    cirq.merge_single_qubit_gates_into_phxz(ret_circuit)
+    cirq.DropEmptyMoments().optimize_circuit(circuit=ret_circuit)
+    ret_circuit = _flatten_circuit(ret_circuit)
+
+    for i, qubit in enumerate([item for sublist in qubit_pairs for item in sublist]):
+        ret_circuit += cirq.measure(qubit, key="q{}".format(i))
+
+    cirq.SynchronizeTerminalMeasurements().optimize_circuit(circuit=ret_circuit)
+    logging.debug(
+        f"Generated a new circuit w/ tsym={use_tsym} and depth {len(ret_circuit)}"
+    )
+    return ret_circuit
+
+
+def _engine_sim_workaround(
+    batch: List[cirq.Circuit], n_shots: int, use_engine: bool
+) -> List[cirq.Result]:
+    """Use either simulator or engine."""
+    if use_engine:
+        project_id = os.environ["GOOGLE_CLOUD_PROJECT"]
+        engine = cirq_google.Engine(project_id=project_id)
+        job = engine.run_batch(
+            programs=batch,
+            gate_set=cirq_google.SYC_GATESET,
+            processor_ids=["weber"],
+            repetitions=n_shots,
+        )
+        return job.results()
+
+    sim = cirq.Simulator()
+    return [sim.run(circuit, repetitions=n_shots) for circuit in batch]
+
+
+def run_and_save(
+    n: int,
+    depth: int,
+    n_data: int,
+    batch_size: int,
+    n_shots: int,
+    save_dir: str,
+    use_engine: bool,
+) -> None:
+    """Run and save bitstring data for tsym vs scrambling experiment w/ twocopy.
+
+    Note: uses qubit layouts native to a Sycamore device (Weber).
+
+    Args:
+        n: Number of system qubits to use.
+        depth: Number of tsym or scrambling blocks.
+        n_data: Number of tsym and scrambling circuits to generate.
+        batch_size: The number of circuits to send off for execution in
+            a single payload (Does not affect experimental resluts).
+        save_dir: str or Path to directory where data is saved.
+        use_engine: Whether or not to make use of quantum engine and the
+            weber processor. Note this requires the GOOGLE_CLOUD_PROJECT
+            environment variable to be set, along with the required cloud
+            permissions.
+    """
+    logging.info("Beginning quantum-enhanced circuit generation.")
+    # Choose system pairs so that they are consecutive neighbors.
+    system_pairs = [
```

Comment:
Should we abstract this out to a common file?

Reply:
Also split.